### PR TITLE
ENH: Added missing methods to `np.flatiter`

### LINKED ISSUE #17113: ENH: Add __getitem__ to flatiter type definition
### Reproducing code example:

Run mypy on a simple file

```
import numpy as np

a = np.empty((2,2))
b = a.flat[[0,1]]
```

### Error message:

You will get

```
error: Value of type "flatiter[ndarray]" is not indexable
```

Need to add `__getitem__` to pyi file.

### Numpy/Python version information:

```
1.20.0.dev0+986e533 3.8.3 (default, Jul  2 2020, 16:21:59)
[GCC 7.3.0]
```

### LINKED ISSUE #17114: ENH: flatiter is missing _SupportsArray in type definition
### Reproducing code example:

`flat` fails typing when used with a `ufunc` because `flatiter` is not array-like enough

Run mypy on 

```python
import numpy as np

a = np.empty((2,2))
np.any(a.flat)
```

I think the solution is to have `flatiter` inherit `_SupportsArray` since it does expose `__array__`.

### Error message:

```
demo.py:4: error: No overload variant of "any" matches argument type "flatiter[ndarray]"
demo.py:4: note: Possible overload variants:
demo.py:4: note:     def any(a: Union[bool, int, float, complex, _SupportsArray, Sequence[Any]], axis: None = ..., out: Optional[ndarray] = ..., keepdims: Literal[False] = ...) -> bool_
demo.py:4: note:     def any(a: Union[bool, int, float, complex, _SupportsArray, Sequence[Any]], axis: Union[int, Sequence[int], None] = ..., out: Optional[ndarray] = ..., keepdims: bool = ...) -> Union[bool_, ndarray]
```

### Numpy/Python version information:

```
1.20.0.dev0+986e533 3.8.3 (default, Jul  2 2020, 16:21:59)
[GCC 7.3.0]
```

### PR DESCRIPTION
**Pull request by BvB93**
Closes https://github.com/numpy/numpy/issues/17114 and https://github.com/numpy/numpy/issues/17113.

This pull requests adds annotations to the following three `np.flatiter` methods:

* `__getitem__()`
* `__len__()`
* `__array__()`

### CODE REVIEW

**Comment by BvB93** [comment]
Pinging @bashtage.

File: numpy/tests/typing/fail/flatiter.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+a: "np.flatiter[np.ndarray]"
+supports_array: _SupportsArray
+
+a.base = Any  # E: Property "base" defined in "flatiter" is read-only
+a.coords = Any  # E: Property "coords" defined in "flatiter" is read-only
+a.index = Any  # E: Property "index" defined in "flatiter" is read-only
+a.copy(order='C')  # E: Unexpected keyword argument
+a[np.bool_()]  # E: No overload variant of "__getitem__"
+a[Index()]  # E: No overload variant of "__getitem__"
```
Comment (eric-wieser) [inline numpy/tests/typing/fail/flatiter.py:20]: Shouldn't this be legal? Might be worth referencing an issue if its not.

Comment (BvB93) [reply numpy/tests/typing/fail/flatiter.py:24]: It should be, but the behavior of `flatiter.__getitem__` is very much inconsistent 
with respect to `ndarray.__getitem__` (case in point: https://github.com/numpy/numpy/issues/17175).

Adding a note is probably a good idea though (see https://github.com/numpy/numpy/pull/17180/commits/edcf0ea581546eb8c29ba34b7576f2a8d81b4eec).

File: numpy/__init__.pyi [hunk 2]
```diff
@@ -143,6 +143,14 @@ class _flagsobj:
     def __getitem__(self, key: str) -> bool: ...
     def __setitem__(self, key: str, value: bool) -> None: ...
 
+_ArrayLikeInt = Union[
+    int,
```
Comment (bashtage) [inline numpy/__init__.pyi:147]: Does [`int`, `integer`] include `bool`?

Comment (BvB93) [reply numpy/__init__.pyi:147]: It does as `bool` is a subclass of `int`.

There is, as far as I'm aware, no convenient way of annotating a function which accepts integers but rejects booleans.

**Review by eric-wieser — APPROVED** [review]
Looks safe enough

**Comment by mattip** [comment]
Thanks @BvB93

**Pull request merged**